op-node/rollup/derive/engine_update.go

```diff
@@ -116,6 +124,65 @@ func startPayload(ctx context.Context, eng ExecEngine, fc eth.ForkchoiceState, a
 	}
 }
 
+// makes parallel request to builder and engine to get the payload
+func getPayloadWithBuilderPayload(ctx context.Context, log log.Logger, eng ExecEngine, payloadInfo eth.PayloadInfo, l2head eth.L2BlockRef, builder builder.PayloadBuilder, metrics Metrics) (
+	*eth.ExecutionPayloadEnvelope, *eth.ExecutionPayloadEnvelope, error) {
+	// if builder is not enabled, return early with default path.
+	if !builder.Enabled() {
+		payload, err := eng.GetPayload(ctx, payloadInfo)
+		return payload, nil, err
+	}
+
+	log.Debug("requesting payload from builder", l2head.String(), "payloadInfo", payloadInfo)
+	ctxTimeout, cancel := context.WithTimeout(ctx, builder.Timeout())
+	defer cancel()
+	type result struct {
+		envelope *eth.ExecutionPayloadEnvelope
+		profit   *big.Int
+		error    error
+	}
+
+	ch := make(chan *result, 1)
+
+	// start the payload request to builder api
+	go func() {
+		start := time.Now()
+		payload, profit, err := builder.GetPayload(ctxTimeout, l2head, log)
+		metrics.RecordBuilderRequestTime(time.Since(start))
+		if err != nil {
+			ch <- &result{error: err}
+			return
+		}
+		ch <- &result{envelope: payload, profit: profit}
+	}()
+
+	envelope, err := eng.GetPayload(ctx, payloadInfo)
```

Comment:
why not have both requests be goroutines and use `sync.WaitGroup` to sync the two requests? This way, if `eng.GetPayload` errors but the builder payload does not, we can still use the builder payload

Reply:
I think the current implementation where the `select` waits for the channel is more golang native than using sync.WaitGroup. Also, it would increase the complexity because it needs to read from two channels(or twice from one channel) to get the result from two go routines